### PR TITLE
Cleaned and extended function that extracts datetimes from paths

### LINKED ISSUE #2180: Unify `_get_start_end_date` and `_get_start_end_year` in `local.py`
While working on the ERA5 GRIB CMORizer, I found that we have the two functions [`_get_start_end_date`](https://github.com/ESMValGroup/ESMValCore/blob/4fd8701efd9e618ade0b3ec83a3733274d4f6b25/esmvalcore/local.py#L62C5-L62C24) and [`_get_start_end_year`](https://github.com/ESMValGroup/ESMValCore/blob/4fd8701efd9e618ade0b3ec83a3733274d4f6b25/esmvalcore/local.py#L165C5-L165C24) in `local.py`, which are very similar. I think we can easily get rid of `_get_start_end_year` by simply reformatting the output of `_get_start_end_year`.

I will open a PR for that.

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR

1. cleans the module `local.py` by removing the duplicated code in the function [`_get_start_end_year`](https://github.com/ESMValGroup/ESMValCore/blob/4fd8701efd9e618ade0b3ec83a3733274d4f6b25/esmvalcore/local.py#L165C5-L165C24) and uses [`_get_start_end_date`](https://github.com/ESMValGroup/ESMValCore/blob/4fd8701efd9e618ade0b3ec83a3733274d4f6b25/esmvalcore/local.py#L62C5-L62C24) instead.
2. extended the regex used to derive datetimes from paths. Now the following strings can be parsed correctly:

```py
'tas_A1.20C3M_1.CCSM.atmm.1990-01_cat_1999-12.nc' -> '199001', '199912'
'E5sf00_1M_1940_032.grb' -> '1940', '1940'
'E5sf00_1D_1998-04_167.grb' -> '199804', '199804'
'E5sf00_1H_1986-04-11_167.grb' -> '19860411', '19860411'
'E5sf00_1M_1940-1941_032.grb' -> '1940', '1941'
'E5sf00_1D_1998-01_1999-12_167.grb' -> '199801', '199912'
'E5sf00_1H_2000-01-01_2001-12-31_167.grb -> '20000101', '20011231'
```

The first is used in CMIP3 data, the latter in native ERA5 data in GRIB format. This speeds up many recipes by quite some time, since only the files that are actually relevant for the user are processed.

Note that I did not change any existing test case, so this should be fully backwards-compatible.

Closes #2180
Addresses parts of #1991  

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Review by bouweandela — COMMENTED** [review]
Nice work @schlunma! Just one minor question

File: esmvalcore/local.py [hunk 1]
```diff
@@ -109,12 +143,30 @@ def _get_start_end_date(filename):
...
-        raise ValueError(f'File {filename} dates do not match a recognized '
-                         'pattern and time can not be read from the file')
+        raise ValueError(
+            f"File {file} datetimes do not match a recognized pattern and "
+            f"time coordinate can not be read from the file"
+        )
+
+    # Remove potential '-' characters from datetimes
+    start_date = start_date.replace('-', '')
+    end_date = end_date.replace('-', '')
```
[the remark this answers is not in the document]
Comment (schlunma) [reply esmvalcore/local.py:153]: I tried, but couldn't make it work :disappointed:

[the remark this answers is not in the document]
Comment (bouweandela) [reply esmvalcore/local.py:153]: I have some ideas, but I'll put them into a new pull request so this can be merged.

**Review by bouweandela — APPROVED** [review]

**Review by valeriupredoi — APPROVED** [review]
biutiful, fellas :beers:

**Pull request merged**